### PR TITLE
Convert Mongo to Instrumenter

### PR DESCRIPTION
**Pull request by iNikem**

### CODE REVIEW

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoAttributesExtractor.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      @Nullable Void unused,
+      @Nullable Throwable error) {}
+
+  @Nullable
+  String collectionName(CommandStartedEvent event) {
+    if (event.getCommandName().equals("getMore")) {
+      if (event.getCommand().containsKey("collection")) {
+        BsonValue collectionValue = event.getCommand().get("collection");
+        if (collectionValue.isString()) {
+          return event.getCommand().getString("collection").getValue();
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoAttributesExtractor.java:38]: Nit
```suggestion
          return collectionValue.asString().getValue();
```

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoAttributesExtractor.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      AttributesBuilder attributes,
+      CommandStartedEvent event,
+      @Nullable Void unused,
+      @Nullable Throwable error) {}
+
+  @Nullable
+  String collectionName(CommandStartedEvent event) {
+    if (event.getCommandName().equals("getMore")) {
+      if (event.getCommand().containsKey("collection")) {
+        BsonValue collectionValue = event.getCommand().get("collection");
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoAttributesExtractor.java:36]: Nit: you could avoid double map lookup by not calling `containsKey`:
```suggestion
      BsonValue collectionValue = event.getCommand().get("collection");
      if (collectionValue != null) {
```

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java [hunk 3]
```diff
@@ -10,38 +10,47 @@
...
+      requestMap.put(event.getRequestId(), event);
+    }
   }
 
   @Override
   public void commandSucceeded(CommandSucceededEvent event) {
     Context context = contextMap.remove(event.getRequestId());
-    if (context != null) {
-      tracer.end(context);
+    CommandStartedEvent request = requestMap.get(event.getRequestId());
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java:42]: Shouldn't it be 
```suggestion
    CommandStartedEvent request = requestMap.remove(event.getRequestId());
```
?

Also, WDYT about maintaining a single map containing a `ContextAndRequest` values?

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoDbAttributesExtractor.java [hunk 4]
```diff
@@ -5,99 +5,62 @@
...
-    super(openTelemetry, NetPeerAttributes.INSTANCE);
+  MongoDbAttributesExtractor(int maxNormalizedQueryLength) {
     this.maxNormalizedQueryLength = maxNormalizedQueryLength;
     this.jsonWriterSettings = createJsonWriterSettings(maxNormalizedQueryLength);
   }
 
   @Override
-  protected String getInstrumentationName() {
-    return "io.opentelemetry.mongo-3.1";
+  protected @Nullable String system(CommandStartedEvent event) {
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoDbAttributesExtractor.java:48]:
```suggestion
  protected String system(CommandStartedEvent event) {
```

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoDbAttributesExtractor.java [hunk 5]
```diff
@@ -114,36 +77,33 @@ protected String dbConnectionString(CommandStartedEvent event) {
...
-    } else {
-      return null;
-    }
-  }
-
-  @Override
-  protected String dbStatement(
-      CommandStartedEvent event, BsonDocument command, String sanitizedStatement) {
-    return sanitizedStatement;
+  protected @Nullable String statement(CommandStartedEvent event) {
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/MongoDbAttributesExtractor.java:80]:
```suggestion
  protected String statement(CommandStartedEvent event) {
```

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java [hunk 6]
```diff
@@ -10,38 +10,47 @@
...
+      instrumenter.end(context, request, null, null);
     }
   }
 
   @Override
   public void commandFailed(CommandFailedEvent event) {
     Context context = contextMap.remove(event.getRequestId());
-    if (context != null) {
-      tracer.endExceptionally(context, event.getThrowable());
+    CommandStartedEvent request = requestMap.get(event.getRequestId());
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java:51]: Here too:
```suggestion
    CommandStartedEvent request = requestMap.remove(event.getRequestId());
```

File: instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java (hunk 3 above)
Comment (trask) [reply instrumentation/mongo/mongo-3.1/library/src/main/java/io/opentelemetry/instrumentation/mongo/v3_1/TracingCommandListener.java]: @iNikem this seems like a good idea

**Review by trask — APPROVED** [review]
> WDYT about maintaining a single map containing a `ContextAndRequest` values?

this seems like a good idea

**Pull request merged**